binder/src/main/java/io/grpc/binder/internal/BinderTransportSecurity.java

```diff
@@ -178,9 +181,12 @@ ListenableFuture<Status> checkAuthorization(MethodDescriptor<?, ?> method) {
       boolean useCache = method.isSampledToLocalTracing();
       if (useCache) {
         @Nullable ListenableFuture<Status> authorization = serviceAuthorization.get(serviceName);
-        if (authorization != null) {
+        if (authorization != null && !isFailedOrCancelled(authorization)) {
+          // Authorization check exists and is a pending or successful future (even if for a failed
+          // authorization).
           return authorization;
         }
+        serviceAuthorization.remove(serviceName);
```

Comment:
I think there's also another race condition from #10670. Added a `synchronized{}` block around the map access to make all access atomic.

Switched to removing on a future callback. Wasn't as complex as I had imagined.

Reply:
Can you say more about the other race you are trying to fix? Calling serverPolicyChecker while holding a lock shared with the whole transport isn't going to fly, especially since all existing policies are not actually async.